tools/base/libboard/board_vc/board_vc.c

```diff
@@ -0,0 +1,746 @@
+
+// sysfs paths
+#define SYSFS_BMCFW_VER                     "spi-altera.*.auto/spi_master/spi*/spi*.*/bmcfw_flash_ctrl/bmcfw_version"
+#define SYSFS_MAX10_VER                     "spi-altera.*.auto/spi_master/spi*/spi*.*/max10_version"
+#define SYSFS_PCB_INFO                      "spi-altera.*.auto/spi_master/spi*/spi*.*/pcb_info"
+#define SYSFS_PKVL_POLL_MODE                "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/polling_mode"
+#define SYSFS_PKVL_STATUS                   "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/status"
+#define SYSFS_BS_ID                         "bitstream_id"
+#define SYSFS_PHY_GROUP_INFO                "misc/eth_group*.*"
+#define SYSFS_PHY_GROUP_INFO_DEV            "misc/eth_group*/dev"
+#define SYSFS_EEPROM                        "*i2c*/i2c*/*/eeprom"
+#define SYSFS_NVMEM                         "*i2c*/i2c*/*/nvmem"
+#define SYSFS_PKVL_A_VER                    "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/pkvl_a_version"
+#define SYSFS_PKVL_B_VER                    "spi-altera.*.auto/spi_master/spi*/spi*.*/pkvl/pkvl_b_version"
+
+// driver ioctl id
+#define FPGA_PHY_GROUP_GET_INFO               0xB702
+
+// Read BMC firmware version
+fpga_result read_bmcfw_version(fpga_token token, char *bmcfw_ver, size_t len)
+{
+	fpga_result res                = FPGA_OK;
+	fpga_result resval             = FPGA_OK;
+	uint8_t rev                    = 0;
+	uint32_t var                   = 0;
+	uint32_t size                  = 0;
+	char buf[VER_BUF_SIZE]         = { 0 };
+	fpga_object bmcfw_object;
+
+	if (bmcfw_ver == NULL) {
+		FPGA_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_BMCFW_VER, &bmcfw_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
+		return res;
+	}
+
+	res = fpgaObjectGetSize(bmcfw_object, &size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object size ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	// Return error if input sysfs path length smaller then object size
+	if (len < size) {
+		FPGA_ERR("Invalid Input sysfs path size");
+		resval = FPGA_EXCEPTION;
+		goto out_destroy;
+	}
+
+	res = fpgaObjectRead(bmcfw_object, (uint8_t *)buf, 0, size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy;
+	}
+
+
+	var = strtoul(buf, NULL, VER_BUF_SIZE);
+
+	/* BMC FW version format reading
+	NIOS II Firmware Build 0x0 32 RW[23:0] 24 hFFFFFF Build version of NIOS II Firmware
+	NIOS FW is up e.g. 1.0.1 for first release
+	[31:24] 8hFF Firmware Support Revision - ASCII code
+	0xFF is the default value without NIOS FW, will be changed after NIOS FW is up
+	0x41(A)-For RevA
+	0x42(B)-For RevB
+	0x43(C)-For RevC
+	0x44(D)-For RevD
+	*/
+
+	rev = (var >> 24) & 0xff;
+	if ((rev > 0x40) && (rev < 0x5B)) {// range from 'A' to 'Z'
+		snprintf_s_ciii(bmcfw_ver, len, "%c.%u.%u.%u", (char)rev, (var >> VER_BUF_SIZE) & 0xff, (var >> 8) & 0xff, var & 0xff);
```

Comment:
parsing the version should probably be refactored into its own function

Reply:
fixed
new function to parse firmware